### ISSUE TITLE
Where to start

### ISSUE THREAD

**Issue by iamricard** [Telegram-FFOS/Telegram-JavaScript-Library#1]
I think we should look at the whole

https://github.com/Telegram-FFOS/webogram/tree/master/app/js/lib

Specifically mtproto.js has +2000 lines of code and I think it could be split into some utility files instead of having everything in a big file.

We should also comment the whole project as we go. Tests are also important.

/cc @dannluciano @novia713 @sebasmagri

**Comment by sebasmagri** [comment]
@rcsole the code looks pretty good despite the lack of comments.

I didn't dig on the app too much, but I believe we'd just try to add a manifest and try packaging the app to get rid of the server side code.

Best Regards,

**Comment by iamricard** [comment]
I'll give it a try in an hour and let you know.

**Comment by iamricard** [comment]
@sebasmagri I tried on both the browser and the simulator and here's what happens:

On the web version it works fine (as expected):
![web version](http://i.imgur.com/lcDRyuX.png)

But then when on the firefox os simulator this happens:
![friend list](http://i.imgur.com/poepw2W.png) ![single chat](http://i.imgur.com/VD0TuWu.png)

My guess that since the app has the views designed to work in a big resolution it tries to render it on the right but it can't and that's why the url doesn't make sense to ffos.

So basically the UI needs to be reworked which should be no big deal. Any thoughts?

**Comment by sebasmagri** [comment]
@rcsole I tried with Firefox responsive mode and the UI was working fine.

I believe you need to use a https:// URL though as per the error message, but it won't work due to mixed content issues. There is a hardcoded server list at https://github.com/Telegram-FFOS/webogram/blob/master/app/js/lib/mtproto.js#L893 and all requests are done over plain http as per https://github.com/Telegram-FFOS/webogram/blob/master/app/js/lib/mtproto.js#L1075.

I'm not sure if such servers are upstream (telegram's) servers obtained with https://core.telegram.org/api/datacenter instructions. We could start by forking webogram and trying to make it work https only to be able to use it on FxOS.

**Comment by iamricard** [comment]
I'll take a look at that, and the repo is already forked on the org. Glad to see more people are insterested!

**Edit:** My point about splitting the project is that things like the datacenter list could go somewhere else and the first few lines where it basically just converts to different data types could be separate so the code would be more readable.

And again, tests would be good although I'm not sure if it's viable to write tests.

**Comment by sebasmagri** [comment]
@rcsole Webogram now work in FxOS. There are still some performance issues, but it should make it to the public marketplace soon.

However, I'd like to leave another option here for documentation purposes; as Telegram's official libraries are implemented in Java, it would be possible to compile them to JavaScript using GWT.

**Comment by iamricard** [comment]
Sorry I've been kind of absent, school is keeping me busy. I'll look into that!
